src/NavigationHelper/src/NavigationHelper.cpp

```diff
@@ -0,0 +1,174 @@
+            m_navigationTriggerThread = std::thread();
+        }
+    }
+    return true;
+}
+
+bool NavigationHelper::closeHelper()
+{
+    try
+    {
+        //close threads first
+        if (m_runThreads)
+        {
+            closeThreads();
+        }
+        //close ports
+        if(!m_replanningTriggerPort.isClosed())
+            m_replanningTriggerPort.close();
+    }
+    catch(const std::exception& e)
+    {
+        std::cerr << e.what() << std::endl;
+        return false;
+    }
+    return true;
+}
+
+bool NavigationHelper::init(const yarp::os::Searchable& config, std::deque<bool> &leftInContact, std::deque<bool> &rightInContact )
+{
+    m_leftInContact = &leftInContact;
+    m_rightInContact = &rightInContact;
+
+    m_navigationReplanningDelay = config.check("navigationReplanningDelay", yarp::os::Value(0.9)).asFloat64();
+    m_navigationTriggerLoopRate = config.check("navigationTriggerLoopRate", yarp::os::Value(100)).asInt32();
+    m_publishInfo = config.check("publishNavigationInfo", yarp::os::Value(false)).asBool();
+    if (config.check("plannerMode", yarp::os::Value("manual")).asString() == "navigation")
+    {
+        //if in navigation mode we need this parameter to be true
+        m_publishInfo = true;
+    }
+    if (!m_publishInfo)
+    {   //exit the funnction if the infos are not needed
+        yInfo() << "[NavigationHelper::init] Configuring NavigationHelper without publishing infos on ports ";
+        return true;
+    }
+
+    // format check
+    if (m_navigationTriggerLoopRate<=0)
+    {
+        yError() << "[configure] navigationTriggerLoopRate must be strictly positive, instead is: " << m_navigationTriggerLoopRate;
+        return false;
+    }
+    if (m_navigationReplanningDelay<0)
+    {
+        yError() << "[configure] navigationTriggerLoopRate must be positive, instead is: " << m_navigationReplanningDelay;
+        return false;
+    }
+
+    std::string replanningTriggerPortPortName = m_portPrefix + "/replanning_trigger:o";
+    if (!m_replanningTriggerPort.open(replanningTriggerPortPortName))
+    {
+        yError() << "[NavigationHelper::init] Could not open" << replanningTriggerPortPortName << " port.";
+        return false;
+    }
+
+    m_runThreads = true;
+    m_navigationTriggerThread = std::thread(&NavigationHelper::computeNavigationTrigger, this);
+    return true;
+}
+
+// This thread synchronizes the walking-controller with the navigation stack.
+// Writes on a port a boolean value when to replan the path
+void NavigationHelper::computeNavigationTrigger()
+{
+    if (!m_publishInfo)
+    {   //exit the funnction if the infos are not needed
+        return;
+    }
+    bool trigger = false;   //flag used to fire the wait for sending the navigation replanning trigger
+    yInfo() << "[NavigationHelper::computeNavigationTrigger] Starting Thread";
+    yarp::os::NetworkClock myClock;
+    myClock.open("/clock", "/navigationTriggerClock");
```

Comment:
What if the clock port is not available? I would suggest enabling this connection from configuration file.

Reply:
This is needed for having a timestamp on the message, which is red by the Navigation stack.
If the port is not available no message should be sent

Reply:
Actually since it never gave me issues in simulation I never thought about it.
If it could be an issue I could pass the clock directly to the object creation or use the yarp manager

Reply:
On the robot we don't have the ``/clock`` port. My suggestion would be to add a parameter in the configuration file to enable its use or not. As far as I can see, the only place where it is used is in https://github.com/robotology/walking-controllers/blob/a0855711c42517d8d6f63245b0fe81345d72c725/src/NavigationHelper/src/NavigationHelper.cpp#L160

If the user does not want to use the ``/clock`` port, you could simply substitute that line with a classical delay. If the user instead wants to use it, you might trigger an error if ``myClock.open`` returns ``false``.

Reply:
> This was done to sync it with the simulation (since the RTF is not 1 and it varies), on the real robot it will be the same of the system clock, right?

I think so



> So a simple sleep will do the work. So the parameter should be just for switching between simulation and real robot.

Sounds good!